### PR TITLE
[Cosmos] split proof queries async client

### PR DESCRIPTION
**Pull request by simorenoh**
This is the PR to address the current gap in the cosmos sdk not being split proof

The gone_retry_policy is used exclusively to refresh the partition key range cache and raise the exception immediately in order to repair the document producers and not waste time retrying several times.

One note on these changes is that I still need to remove certain print statements and comments but want to get input first so I can keep logging the right things in case some further iteration is needed.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/azure/cosmos/_routing/aio/routing_map_provider.py [hunk 1]
```diff
@@ -54,7 +54,7 @@ async def get_overlapping_ranges(self, collection_link, partition_key_ranges):
         overlapping partition key ranges.
 
         :param str collection_link: The name of the collection.
-        :param list partition_key_range: List of partition key range.
+        :param list partition_key_ranges: List of partition key range.
```
Comment (annatisch) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_routing/aio/routing_map_provider.py:57]: This looks like it was public API?

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_retry_utility_async.py [hunk 2]
```diff
@@ -100,6 +103,8 @@ async def ExecuteAsync(client, global_endpoint_manager, function, *args, **kwarg
                 and e.sub_status == SubStatusCodes.READ_SESSION_NOTAVAILABLE
             ):
                 retry_policy = sessionRetry_policy
+            elif exceptions.partition_range_is_gone(e):
+                retry_policy = partition_key_range_gone_retry_policy
```
Comment (annatisch) [inline sdk/cosmos/azure-cosmos/azure/cosmos/aio/_retry_utility_async.py:107]: This 'gone' naming isn't part of the public API, right?

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_retry_utility_async.py [hunk 3]
```diff
@@ -35,6 +35,7 @@
 from .. import _resource_throttle_retry_policy
 from .. import _default_retry_policy
 from .. import _session_retry_policy
+from .. import _gone_retry_policy
```
Comment (annatisch) [inline sdk/cosmos/azure-cosmos/azure/cosmos/aio/_retry_utility_async.py:38]: Will these changes be included in sync too?
If so - would it be breaking?

File: sdk/cosmos/azure-cosmos/azure/cosmos/_execution_context/aio/document_producer.py [hunk 4]
```diff
@@ -58,13 +58,18 @@ def __init__(self, partition_key_target_range, client, collection_link, query, d
...
         async def fetch_fn(options):
-            return await self._client.QueryFeed(path, collection_id, query, options, partition_key_target_range["id"])
+            try:
+                return await self._client.QueryFeed(path, collection_id, query, options,
+                                                    partition_key_target_range["id"])
+            except exceptions.CosmosHttpResponseError as e:
+                if e.status_code == http_constants.StatusCodes.GONE:
+                    # 410 within fetch function, raise to properly handle within execution context
+                    raise
+                return await self._client.QueryFeed(path, collection_id, query, options,
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_execution_context/aio/document_producer.py:68]: you're absolutely right, the final version of the document producer shouldn't even need these changes, already fixed this in the sync version but missed it here, thanks!

File: sdk/cosmos/azure-cosmos/azure/cosmos/_routing/aio/routing_map_provider.py (hunk 1 above)
Comment (simorenoh) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_routing/aio/routing_map_provider.py:57]: the routing module is already supposed to be private, so this method does not/ should not get used right?

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_retry_utility_async.py (hunk 3 above)
Comment (simorenoh) [reply sdk/cosmos/azure-cosmos/azure/cosmos/aio/_retry_utility_async.py:38]: they are also included in the sync and it shouldn't be breaking - we are fixing a gap that was present by adding this functionality. moreover, the user doesn't get to interact with the retry utility or the retry policies.
the sync changes are here if you're curious, I'll be making changes to this async fixes to have them be closer in pairity, the big exception being the async heap: https://github.com/Azure/azure-sdk-for-python/pull/22237/files

File: sdk/cosmos/azure-cosmos/azure/cosmos/_routing/aio/routing_map_provider.py (hunk 1 above)
Comment (annatisch) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_routing/aio/routing_map_provider.py:57]: I think before we go GA we should make most of these files in `aio` directory private. It was a big mistake to make them public in the sync version.... and I don't think we gain anything by repeating it here....

File: sdk/cosmos/azure-cosmos/azure/cosmos/exceptions.py [hunk 5]
```diff
@@ -73,3 +73,10 @@ def __init__(self, **kwargs):
         self.response = None
         self.history = None
         super(CosmosClientTimeoutError, self).__init__(message, **kwargs)
+
+
+def partition_range_is_gone(e):
```
Comment (annatisch) [inline sdk/cosmos/azure-cosmos/azure/cosmos/exceptions.py:78]: This function should be private (i.e. underscore prefix)

File: sdk/cosmos/azure-cosmos/azure/cosmos/_routing/aio/routing_map_provider.py (hunk 1 above)
Comment (simorenoh) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_routing/aio/routing_map_provider.py:57]: I'll add this to the list of issues to take care of before GA then, thanks Anna!

**Review by kushagraThapar — CHANGES_REQUESTED** [review]

File: sdk/cosmos/azure-cosmos/CHANGELOG.md [hunk 6]
```diff
@@ -1,18 +1,25 @@
...
+#### Features Added
+- Added support for split-proof queries for the async client
+
 ### 4.3.0b2 (2022-01-25)
 
 This version and all future versions will require Python 3.6+. Python 2.7 is no longer supported.
 We will also be removing support for Python 3.6 and will only support Python 3.7+ starting December 2022.
 
 #### Features Added
+- Added support for split-proof queries for the async client
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure-cosmos/CHANGELOG.md:14]: oops, will take this off, thanks!

**Review by kushagraThapar — APPROVED** [review]

**Pull request merged**